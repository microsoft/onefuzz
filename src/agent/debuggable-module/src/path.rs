// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use std::ffi::OsStr;
use std::fmt;
use std::path::{Path, PathBuf};

use anyhow::{bail, Result};

/// Path to a file. Guaranteed UTF-8.
#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct FilePath(String);

impl FilePath {
    pub fn new(path: impl Into<String>) -> Result<Self> {
        let path = path.into();

        if Path::new(&path).file_name().is_none() {
            bail!("module path has no file name");
        }

        if Path::new(&path).file_stem().is_none() {
            bail!("module path has no file stem");
        }

        Ok(Self(path))
    }

    pub fn with_extension(&self, extension: impl AsRef<str>) -> Self {
        let path = self
            .as_path()
            .with_extension(extension.as_ref())
            .to_string_lossy()
            .into_owned();

        Self(path)
    }

    pub fn as_path(&self) -> &Path {
        Path::new(&self.0)
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn file_name(&self) -> &str {
        // Unwraps checked by ctor.
        Path::new(&self.0).file_name().unwrap().to_str().unwrap()
    }

    pub fn directory(&self) -> &str {
        // Unwraps checked by ctor.
        Path::new(&self.0).parent().unwrap().to_str().unwrap()
    }

    pub fn base_name(&self) -> &str {
        // Unwraps checked by ctor.
        Path::new(&self.0).file_stem().unwrap().to_str().unwrap()
    }
}

impl From<FilePath> for String {
    fn from(path: FilePath) -> Self {
        path.0
    }
}
impl From<FilePath> for PathBuf {
    fn from(path: FilePath) -> Self {
        path.0.into()
    }
}

impl AsRef<str> for FilePath {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl AsRef<OsStr> for FilePath {
    fn as_ref(&self) -> &OsStr {
        self.as_str().as_ref()
    }
}

impl AsRef<Path> for FilePath {
    fn as_ref(&self) -> &Path {
        self.as_path()
    }
}

impl fmt::Display for FilePath {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}
