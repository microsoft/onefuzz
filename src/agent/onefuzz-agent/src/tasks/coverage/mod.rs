// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#[cfg(any(target_os = "linux", target_os = "windows"))]
pub mod generic;
#[cfg(any(target_os = "linux", target_os = "windows"))]
pub mod libfuzzer_coverage;
#[cfg(any(target_os = "linux", target_os = "windows"))]
pub mod recorder;
pub mod total;
