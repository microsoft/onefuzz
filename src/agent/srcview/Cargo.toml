[package]
name = "srcview"
version = "0.1.2"
edition = "2018"
license = "MIT"

[features]
# Feature to gate tests that depend on binary artifacts, should be removed when
# #1143 is resolved.
binary-tests = []

[dependencies]
log = "0.4"
nom = "7"
pdb = "0.8"
regex = "1"
serde = { version = "1", features = ["derive"] }
quick-xml = "0.30"
anyhow = "1.0"
env_logger = "0.10"
clap = { version = "4.3.0", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.31.0", features = ["glob", "json"] }
coverage = { path = "../coverage" }
cc = "1.0"
tempfile = "3.7.0"
dunce = "1.0"
