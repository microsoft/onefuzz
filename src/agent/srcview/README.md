# srcview

A library for mapping module+offset to source:line. Note that you'll get
significantly better results if you use instruction level coverage as
opposed to branch. Alternatively you're welcome to post-process branch
coverage into instruction, but this project does not assist with that.

## Docs

`cargo doc --open`

## Usage

See [`examples/dump_cobertura.rs`](examples/dump_cobertura.rs).

This can be run with `cargo run --example dump_cobertura res\example.pdb res\example.txt`.

The biggest challenge when making this work is likely getting the absolute PDB
paths to match relative paths inside your repo. To make this a bit easier, you
can dump the PDB paths with: `cargo run --example dump_paths res\example.pdb`

## ADO Integration

See [`azure-pipelines.yml`](azure-pipelines.yml).

## VSCode Integration

Install [Coverage Gutters](https://marketplace.visualstudio.com/items?itemName=ryanluker.vscode-coverage-gutters),
then place a file named `coverage.xml` at a location such that the relative
file paths match your repo. Note that this is a file you should generate
yourself, and that the `coverage.xml` included in this repo will probably not
help you. From there, navigate to any file you expect to see coverage in. If
the paths and coverage file correctly line up, you should see red or green bars
next to the source lines.