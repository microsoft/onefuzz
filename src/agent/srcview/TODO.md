# TODO
- add trace/info/warn
- flesh out modoff parser error
- consider using Cow to reduce memory usage
- consider mixed case module names?
- redo xml generation to not be hand-rolled