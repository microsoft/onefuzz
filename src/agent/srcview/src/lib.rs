// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

//! # srcview
//!
//! srcview is a crate for converting module+offset (modoff) coverage traces to source and line
//! number such that they can be visualized in an editor web frontend. It's job is split into
//! two parts: first creating a `SrcView` from debug info, then turning that `SrcView` and coverage into
//! a `Report`.
//!
//! ## SrcView
//!
//! There are two fundamental datatypes the crate defines, `ModOff` and `SrcLine`. `ModOff` is a module
//! name + offset from it's base, and SrcLine is an absolute path and line number. From a high
//! level, srcview converts modoff to srcline using debug info (PDBs only presently), then collects
//! and formats that debug info into a report. A `SrcView` is a collection of debug info (PdbCache)
//! mappings, and each debug info exposes four mappings:
//!  - ModOff to SrcLine (1:n)
//!  - Path to all valid SrcLines (1:n)
//!  - Path to all Symbols in that file (1:n)
//!  - Symbol to all valid SrcLines (1:n)
//!
//! A SrcView is a queryable collection of debug information, it _does not_ contain any coverage
//! information directly. It is simply the data extracted from the relevant debug structures in a
//! conviently queryable form.
//!
//! ## SrcView Examples
//!
//! ```text
//!              module                    offset
//!
//! ┌─────────┐
//! │         │           ┌──────────────┐
//! │ SrcView │ ─────┬───►│PdbCache - foo│
//! │         │      │    └──────────────┘         ┌──────────────────────────┐
//! └─────────┘      │                         ┌──►│SrcLine - z:\src\fizz.c:41│
//!                  │    ┌──────────────┐     │   │SrcLine - z:\src\fizz.c:42│
//!                  ├───►│PdbCache - bar├─────┤   └──────────────────────────┘
//!                  │    └──────────────┘     │
//!                  │                         │   ┌──────────────────────────┐
//!                  │    ┌──────────────┐     └──►│SrcLine - z:\src\quux.c:65│
//!                  └───►│PdbCache - baz│         └──────────────────────────┘
//!                       └──────────────┘
//! ```
//!
//! For example, bar+1234 might map to z:\src\quux.c:65.
//!
//! Absolute path to all valid SrcLine or Symbols in that path:
//! ```text
//!                <all>                    path
//! ┌─────────┐
//! │         │           ┌──────────────┐
//! │ SrcView │ ─────┬───►│PdbCache - foo│
//! │         │      │    └──────────────┘         ┌──────────────────────────┐
//! └─────────┘      │                         ┌──►│SrcLine - z:\src\fizz.c:41│
//!                  │    ┌──────────────┐     │   │SrcLine - z:\src\fizz.c:42│
//!                  ├───►│PdbCache - bar├─────┤   └──────────────────────────┘
//!                  │    └──────────────┘     │
//!                  │                         │   ┌──────────────────────────┐
//!                  │    ┌──────────────┐     └──►│SrcLine - z:\src\quux.c:65│
//!                  └───►│PdbCache - baz├──┐      └──────────────────────────┘
//!                       └──────────────┘  │
//!                                         │      ┌──────────────────────────┐
//!                                         └─────►│SrcLine - z:\src\fizz.c:43│
//!                                                └──────────────────────────┘
//! ```
//!
//! For example if we were asking for SrcLines, z:\src\fizz.c would iterate over all
//! pdbcache's and return a list of z:\src\fizz.c:41, z:\src\fizz.c:42, z:\src.fizz.c:65.
//! If we were asking for a list of symbols for a path, we might get back
//! FunctionOne, FunctionTwo, SomeGlobal, etc.
//!
//! Finally, we can query for symbol (e.g. `foo!FunctionOne`) to `SrcLine`. This functions
//! exactly like the others and will return a collection of `SrcLine`.
//!
//! ## Report
//!
//! Once we have a `SrcView`, we can combine that with the coverage set (i.e. `&[ModOff]`) to create
//! a `Report`. A `Report` is fundamentally the same info as a SrcView, except keyed around file paths
//! and with some statistics computed from the coverage info.
//!
//! Additionally, `Report` handles some of the messier parts:
//!  - Using regex's to filter `SrcView` contents, this lets you exclude things like the CRT from your
//!    output reports.
//!  - Using regex's to fixup input paths. All paths from PDBs are absolute, but most coverage
//!    visualization wants them to match a directory path. For example, if you have a git repo in
//!    ADO that is named 'test', it might be built on a build machine under 'z:\src\test`. To get
//!    the paths to match, we need to filter off 'z:\src\test'.
//!  - The actually emission of the coverage report itself. Right now this is only in Cobertura, but
//!    its not hard to imagine other formats being added.
//!  - Compute the coverage statistics on directories
//!
//! `Report` is significantly messier than `SrcView` and as of writing this I expect there to still be bugs.
//!
mod modoff;
mod pdbcache;
mod report;
mod srcline;
mod srcview;

pub use self::srcview::SrcView;
pub use modoff::{ModOff, ModOffParseError};
pub use pdbcache::PdbCache;
pub use report::Report;
pub use srcline::SrcLine;
