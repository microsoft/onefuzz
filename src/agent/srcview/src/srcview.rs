// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};

use anyhow::Result;
use serde::{Deserialize, Serialize};

use crate::{ModOff, PdbCache, SrcLine};

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct SrcView(BTreeMap<String, PdbCache>);

/// A SrcView is a collection of zero or more PdbCaches for easy querying. It stores all
/// the mapping information from the PDBs. It does _not_ contain any coverage information.
impl SrcView {
    /// Create's a new SrcView
    pub fn new() -> Self {
        Self::default()
    }

    /// Insert a new pdb into the SrcView, returning any previous pdb info that you're
    /// replacing. In most cases this return value can be ignored.
    ///
    /// # Arguments
    ///
    /// * `module` - Module name to store the PDB info as
    /// * `pdb` - Path to PDB
    ///
    /// # Errors
    ///
    ///  If the PDB at the provided path cannot be parsed or contains otherwise unexpected data.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use srcview::SrcView;
    ///
    /// let mut sv = SrcView::new();
    ///
    /// // Map the contents of 'example.pdb' to the module name 'example.exe'
    /// sv.insert("example.exe", r"z:\src\example.pdb").unwrap();
    ///
    /// // you can now query sv for info from example.exe...
    /// ```
    pub fn insert<P: AsRef<Path>>(&mut self, module: &str, pdb: P) -> Result<Option<PdbCache>> {
        let cache = PdbCache::new(pdb)?;
        Ok(self.0.insert(module.to_owned(), cache))
    }

    /// Insert a new pdb into the SrcView only if the `pdb` path is not in the SrcView already,
    /// returning a  [Result] indicating the success of the insert, if any was necessary.
    /// If the [Result] is [Ok], the contained bool indicates whether a value was inserted.
    ///
    /// # Arguments
    ///
    /// * `module` - Module name to store the PDB info as
    /// * `pdb` - Path to PDB
    ///
    /// # Errors
    ///
    /// If the PDB at the provided path has not been inserted already **and**
    /// cannot be parsed or contains otherwise unexpected data.
    ///
    /// # Example
    ///
    /// ```ignore
    /// use srcview::SrcView;
    ///
    /// let mut sv = SrcView::new();
    /// let modoffs = get_coverage();
    ///
    /// // Map each modoff to a PDB name/path and make sure it's in the SrcView
    /// for modoff in modoffs {
    ///     let module_name =  mod_name_from_modoff(modoff);
    ///     let res = sv.try_insert(module_name, format!("~/pdbs/{module_name}.pdb"));
    ///
    ///     if let Ok(inserted) = res {
    ///         println!("PDB was inserted: {inserted}");
    ///     }
    /// }
    /// ```
    pub fn try_insert<P: AsRef<Path>>(&mut self, module: &str, pdb: P) -> Result<bool> {
        if self.0.contains_key(&module.to_owned()) {
            Ok(false)
        } else {
            self.insert(module, pdb).map(|_| true)
        }
    }

    /// Resolve a modoff to SrcLine, if one exists
    ///
    /// # Arguments
    ///
    /// * `modoff` - Reference to a ModOff you'd like to resolve
    ///
    /// # Example
    ///
    /// ```no_run
    /// use srcview::{ModOff, SrcView};
    ///
    /// let mut sv = SrcView::new();
    ///
    /// // Map the contents of 'example.pdb' to the module name 'example.exe'
    /// sv.insert("example.exe", r"z:\src\example.pdb").unwrap();
    ///
    /// let modoff = ModOff::new("example.exe", 0x4141);
    ///
    /// if let Some(srcline) = sv.modoff(&modoff) {
    ///     println!("Resolved {} to {}", modoff, srcline);
    /// }
    /// ```
    pub fn modoff(&self, modoff: &ModOff) -> Option<SrcLine> {
        match self.0.get(&modoff.module) {
            Some(cache) => cache.offset(&modoff.offset).cloned(),
            None => None,
        }
    }

    /// Resolve a symbol (e.g. module!name) to its possible SrcLines, if such a symbol
    /// exists
    ///
    /// # Arguments
    ///
    /// * `sym` - A symbol name you'd like to query in the form of `<module>!<name>`
    ///
    /// # Example
    ///
    /// ```no_run
    /// use srcview::SrcView;
    ///
    /// let mut sv = SrcView::new();
    ///
    /// // Map the contents of 'example.pdb' to the module name 'example.exe'
    /// sv.insert("example.exe", r"z:\src\example.pdb").unwrap();
    ///
    /// if let Some(srclines) = sv.symbol("example.exe!main") {
    ///     println!("example.exe!main has the following valid source lines:");
    ///     for line in srclines {
    ///         println!(" - {}", line);
    ///     }
    /// };
    /// ```
    pub fn symbol(&self, sym: &str) -> Option<impl Iterator<Item = &SrcLine>> {
        let split: Vec<&str> = sym.split('!').collect();

        if split.len() < 2 {
            return None;
        }

        let module = split[0];
        let name: String = split[1..].join("!");

        match self.0.get(module) {
            Some(cache) => cache.symbol(&name),
            None => None,
        }
    }

    /// Resolve a path to its possible SrcLines, if such a path exists
    ///
    /// # Arguments
    ///
    /// * `path` - An absolute path that possibly matches one from the debug info
    ///
    /// # Example
    ///
    /// ```no_run
    /// use srcview::SrcView;
    ///
    /// let mut sv = SrcView::new();
    ///
    /// // Map the contents of 'example.pdb' to the module name 'example.exe'
    /// sv.insert("example.exe", r"z:\src\example.pdb").unwrap();
    ///
    /// if let Some(srclines) = sv.path_lines(r"z:\src\example.c") {
    ///     println!("example.c has the following valid source lines:");
    ///     for line in srclines {
    ///         println!(" - {}", line);
    ///     }
    /// }
    /// ```
    pub fn path_lines<P: AsRef<Path>>(&self, path: P) -> Option<impl Iterator<Item = usize>> {
        // we want to unique the lines in use across all loaded pdbs
        let mut r: BTreeSet<usize> = BTreeSet::new();

        for cache in self.0.values() {
            if let Some(lines) = cache.path_lines(path.as_ref()) {
                for line in lines {
                    r.insert(*line);
                }
            }
        }

        if r.is_empty() {
            return None;
        }

        // convert to a vec
        let mut v: Vec<usize> = r.into_iter().collect();

        // lists of line numbers are nicer when theyre sorted...
        v.sort_unstable();

        Some(v.into_iter())
    }

    /// Resolve a path to its possible symbols, if such a path exists
    ///
    /// # Arguments
    ///
    /// * `path` - An absolute path that possibly matches one from the debug info
    ///
    /// # Example
    ///
    /// ```no_run
    /// use srcview::SrcView;
    ///
    /// let mut sv = SrcView::new();
    ///
    /// // Map the contents of 'example.pdb' to the module name 'example.exe'
    /// sv.insert("example.exe", r"z:\src\example.pdb").unwrap();
    ///
    /// if let Some(symbols) = sv.path_symbols(r"z:\src\example.c") {
    ///     println!("example.c has the following valid symbols:");
    ///     for line in symbols {
    ///         println!(" - {}", line);
    ///     }
    /// }
    /// ```
    pub fn path_symbols<P: AsRef<Path>>(&self, path: P) -> Option<impl Iterator<Item = String>> {
        // we want to unique the lines in use across all loaded pdbs
        let mut r: BTreeSet<String> = BTreeSet::new();

        for (module, cache) in self.0.iter() {
            if let Some(symbols) = cache.path_symbols(path.as_ref()) {
                for sym in symbols {
                    r.insert(format!("{module}!{sym}"));
                }
            }
        }

        if r.is_empty() {
            return None;
        }

        // convert to a vec
        let mut v: Vec<String> = r.into_iter().collect();

        // lists of symbols are nicer when theyre sorted...
        v.sort();

        Some(v.into_iter())
    }

    /// Returns an iterator over all paths in the SrcView
    ///
    /// # Example
    ///
    /// ```no_run
    /// use srcview::SrcView;
    ///
    /// let mut sv = SrcView::new();
    ///
    /// // Map the contents of 'example.pdb' to the module name 'example.exe'
    /// sv.insert("example.exe", r"z:\src\example.pdb").unwrap();
    ///
    /// println!("paths in example.pdb:");
    ///
    /// for path in sv.paths() {
    ///     println!(" - {}", path.display());
    /// }
    /// ```
    pub fn paths(&self) -> impl Iterator<Item = &PathBuf> {
        let mut r: BTreeSet<&PathBuf> = BTreeSet::new();

        for cache in self.0.values() {
            for pb in cache.paths() {
                r.insert(pb);
            }
        }

        r.into_iter()
    }
}
