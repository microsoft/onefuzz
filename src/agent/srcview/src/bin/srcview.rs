// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use anyhow::{format_err, Context, Result};
use srcview::{ModOff, Report, SrcLine, SrcView};
use std::fs::{self};
use std::io::{stdout, Write};
use std::path::{Path, PathBuf};
use structopt::StructOpt;

#[derive(StructOpt, Debug)]
enum Opt {
    Srcloc(SrcLocOpt),
    PdbPaths(PdbPathsOpt),
    Cobertura(CoberturaOpt),
    /// Print 3rd-party license information
    Licenses,
}

/// Print the file paths in the provided PDB
#[derive(StructOpt, Debug)]
struct PdbPathsOpt {
    pdb_path: PathBuf,
}

/// Print modoffset file with file and source lines
#[derive(StructOpt, Debug)]
struct SrcLocOpt {
    pdb_path: PathBuf,
    modoff_path: PathBuf,
    #[structopt(long)]
    module_name: Option<String>,
}

/// Generate a Cobertura XML coverage report
///
/// Example:
///   srcview cobertura ./res/example.pdb res/example.txt
///             --include-regex "E:\\\\1f\\\\coverage\\\\"
///             --filter-regex "E:\\\\1f\\\\coverage\\\\"
///             --module-name example.exe
///
/// In this example, only files that live in E:\1f\coverage are included and
/// E:\1f\coverage is removed from the filenames in the resulting XML report.
#[derive(StructOpt, Debug)]
struct CoberturaOpt {
    pdb_path: PathBuf,
    modoff_path: PathBuf,
    #[structopt(long)]
    module_name: Option<String>,

    /// regular expression that will be applied against the file paths from the
    /// srcview
    #[structopt(long)]
    include_regex: Option<String>,

    /// search and replace regular expression that is applied to all file
    /// paths that will appear in the output report
    #[structopt(long)]
    filter_regex: Option<String>,
}

fn main() -> Result<()> {
    env_logger::init();

    let opt = Opt::from_args();

    match opt {
        Opt::Srcloc(opts) => srcloc(opts)?,
        Opt::PdbPaths(opts) => pdb_paths(opts)?,
        Opt::Cobertura(opts) => cobertura(opts)?,
        Opt::Licenses => licenses()?,
    };

    Ok(())
}

fn licenses() -> Result<()> {
    stdout().write_all(include_bytes!("../../../data/licenses.json"))?;
    Ok(())
}

fn add_common_extensions(srcview: &mut SrcView, pdb_path: &Path) -> Result<()> {
    let pdb_file_name = pdb_path.file_name().ok_or_else(|| {
        format_err!(
            "unable to identify file name from path: {}",
            pdb_path.display()
        )
    })?;

    let stem = Path::new(pdb_file_name)
        .file_stem()
        .ok_or_else(|| {
            format_err!(
                "unable to identify file stem from path: {}",
                pdb_path.display()
            )
        })?
        .to_string_lossy();

    // add module without extension
    srcview.insert(&stem, &pdb_path)?;
    // add common module extensions
    for ext in ["sys", "exe", "dll"] {
        srcview.insert(&format!("{}.{}", stem, ext), pdb_path)?;
    }
    Ok(())
}

fn srcloc(opts: SrcLocOpt) -> Result<()> {
    let modoff_data = fs::read_to_string(&opts.modoff_path)
        .with_context(|| format!("unable to read modoff_path: {}", opts.modoff_path.display()))?;
    let modoffs = ModOff::parse(&modoff_data)?;
    let mut srcview = SrcView::new();

    if let Some(module_name) = &opts.module_name {
        srcview.insert(module_name, &opts.pdb_path)?;
    } else {
        add_common_extensions(&mut srcview, &opts.pdb_path)?;
    }

    for modoff in &modoffs {
        print!(" +{:04x} ", modoff.offset);
        match srcview.modoff(modoff) {
            Some(srcloc) => println!("{}", srcloc),
            None => println!(),
        }
    }

    Ok(())
}

fn pdb_paths(opts: PdbPathsOpt) -> Result<()> {
    let mut srcview = SrcView::new();
    srcview.insert(&*opts.pdb_path.to_string_lossy(), &opts.pdb_path)?;

    for path in srcview.paths() {
        println!("{}", path.display());
    }
    Ok(())
}

fn cobertura(opts: CoberturaOpt) -> Result<()> {
    // read our modoff file and parse it to a vector
    let modoff_data = fs::read_to_string(&opts.modoff_path)?;
    let modoffs = ModOff::parse(&modoff_data)?;

    // create our new SrcView and insert our only pdb into it
    // we don't know what the modoff module will be, so create a mapping from
    // all likely names to the pdb
    let mut srcview = SrcView::new();

    if let Some(module_name) = &opts.module_name {
        srcview.insert(module_name, &opts.pdb_path)?;
    } else {
        add_common_extensions(&mut srcview, &opts.pdb_path)?;
    }

    // Convert our ModOffs to SrcLine so we can draw it
    let coverage: Vec<SrcLine> = modoffs
        .into_iter()
        .filter_map(|m| srcview.modoff(&m))
        .collect();

    // Generate our report, filtering on our example path
    let r = Report::new(&coverage, &srcview, opts.include_regex.as_deref())?;

    // Format it as cobertura and display it
    let formatted = r.cobertura(opts.filter_regex.as_deref())?;
    println!("{}", formatted);

    Ok(())
}
