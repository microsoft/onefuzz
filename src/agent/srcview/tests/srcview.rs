// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

// tests depends on example.pdb
// $ sha256sum example.pdb
// ecc4214d687c97e9c8afd0c84b4b75383eaa0a237f8a8ca5049478f63b2c98b9  example.pdb

use std::env;
use std::path::PathBuf;

use srcview::{ModOff, SrcLine, SrcView};

fn test_srcview() -> SrcView {
    let root = env::var("CARGO_MANIFEST_DIR").unwrap();
    let pdb_path: PathBuf = [&root, "res", "example.pdb"].iter().collect();

    let mut srcview = SrcView::new();
    srcview.insert("example.exe", pdb_path).unwrap();

    srcview
}

#[test]
#[cfg_attr(not(feature = "binary-tests"), ignore)]
fn modoff() {
    let srcview = test_srcview();

    let good_modoff = ModOff::new("example.exe", 0x6f70);
    assert_eq!(
        srcview.modoff(&good_modoff),
        Some(SrcLine::new("E:\\1f\\coverage\\example\\example.c", 3))
    );

    let bad_offset = ModOff::new("example.exe", 0x4141);
    assert_eq!(srcview.modoff(&bad_offset), None);

    let bad_module = ModOff::new("foo.exe", 0x4141);
    assert_eq!(srcview.modoff(&bad_module), None);
}

#[test]
#[cfg_attr(not(feature = "binary-tests"), ignore)]
fn symbol() {
    let srcview = test_srcview();

    let good: Vec<&SrcLine> = srcview.symbol("example.exe!main").unwrap().collect();

    assert_eq!(
        good,
        vec![
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 3),
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 4),
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 5),
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 6),
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 7),
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 10),
            &SrcLine::new("E:\\1f\\coverage\\example\\example.c", 11),
        ]
    );

    assert!(srcview.symbol("dosenotexist").is_none());
}

#[test]
#[cfg_attr(not(feature = "binary-tests"), ignore)]
fn path() {
    let srcview = test_srcview();

    let good: Vec<usize> = srcview
        .path_lines("E:\\1f\\coverage\\example\\example.c")
        .unwrap()
        .collect();

    assert_eq!(good, vec![3, 4, 5, 6, 7, 10, 11]);

    assert!(srcview.path_lines("z:\\does\\not\\exist.c").is_none());
}

#[test]
#[cfg(target_os = "windows")]
fn windows_snapshot_tests() {
    use coverage::{binary::DebugInfoCache, AllowList};
    use std::time::Duration;

    insta::glob!("windows", "*.cpp", |path| {
        let file_name = path.file_name().unwrap();

        // locate appropriate compiler
        let mut cl_exe =
            cc::windows_registry::find("x86_64-pc-windows-msvc", "cl.exe").expect("finding cl.exe");

        // path will have \\?\ prefix, cl.exe doesn't like it
        let input_path = dunce::canonicalize(path).unwrap();

        // directory to compile into
        let build_in = tempfile::tempdir().expect("creating tempdir");

        let output = cl_exe
            .arg("/EHsc")
            .arg("/Zi")
            .arg("/O2")
            .arg(&input_path)
            .current_dir(build_in.path())
            .spawn()
            .expect("launching compiler")
            .wait_with_output()
            .expect("waiting for compiler to finish");

        assert!(output.status.success(), "cl.exe failed: {:?}", output);

        let exe_name = {
            let mut cwd = build_in.path().to_path_buf();
            cwd.push(file_name);
            cwd.set_extension("exe");
            cwd
        };

        // filter to just the input test file:
        let source_filter = AllowList::parse(&input_path.to_string_lossy()).unwrap();

        let exe_cmd = std::process::Command::new(&exe_name);
        let recorded = coverage::CoverageRecorder::new(exe_cmd)
            .timeout(Duration::from_secs(120))
            .debuginfo_cache(DebugInfoCache::new(source_filter))
            .record()
            .unwrap();

        // generate information with srcview
        let target_file_name = exe_name.file_name().unwrap().to_string_lossy().into_owned();
        let mut srcview = SrcView::new();
        srcview
            .insert(&target_file_name, &exe_name.with_extension("pdb"))
            .unwrap();

        let mut srclines_hit: Vec<SrcLine> = Vec::new();
        for (module, coverage) in &recorded.coverage.modules {
            let module_name = module.file_name();
            for (offset, count) in &coverage.offsets {
                if count.0 > 0 {
                    if let Some(line) = srcview.modoff(&ModOff {
                        module: module_name.to_string(),
                        offset: offset.0 as usize,
                    }) {
                        srclines_hit.push(line);
                    };
                }
            }
        }

        // apply filter to make output stable, replace prefix of path with "…"
        insta::with_settings!({filters => vec![(r"[A-Z]:.*windows", "…")]}, {
            insta::assert_json_snapshot!(srclines_hit);
        });
    });
}
