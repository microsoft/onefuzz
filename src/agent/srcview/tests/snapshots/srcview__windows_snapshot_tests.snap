---
source: srcview/tests/srcview.rs
expression: srclines_hit
input_file: srcview/tests/windows/inlinee.cpp
---
[
  {
    "path": "D:\\git\\onefuzz\\src\\agent\\srcview\\tests\\windows\\inlinee.cpp",
    "line": 6
  }
]
